### PR TITLE
[Notifier] Added Iterable notifier bridge

### PR DESCRIPTION
**Pull request by balazscsaba2006**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | ~
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/15438

[Notifier] Added integration with [Iterable](https://iterable.com/) for the Notifier component.

### CODE REVIEW

File: src/Symfony/Component/Notifier/Bridge/Iterable/IterableTransport.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @author Balázs Csaba <csaba.balazs@lingoda.com>
+ */
+final class IterableTransport extends AbstractTransport
+{
+    protected const HOST = 'api.iterable.com/api/push/target';
+    private const RESPONSE_CODE_SUCCESS = 'Success';
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iterable/IterableTransport.php]: This constant can be removed

File: src/Symfony/Component/Notifier/Bridge/Iterable/IterableTransport.php [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    private $apiKey;
+
+    /** @var string|null */
+    private $campaignId;
+
+    public function __construct(string $apiKey, ?string $campaignId, ?HttpClientInterface $client = null, ?EventDispatcherInterface $dispatcher = null)
+    {
+        $this->apiKey = $apiKey;
+        $this->campaignId = $campaignId;
+        $this->client = $client;
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iterable/IterableTransport.php]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/Iterable/README.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# Iterable Notifier
+
+Provides [Iterable](https://iterable.com/) integration for Symfony Notifier.
+
+## DSN example
+
+```
+ITERABLE_DSN=iterable://API_KEY@default?campaign_id=ID
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iterable/README.md]:
```suggestion
ITERABLE_DSN=iterable://API_KEY@default?campaign_id=CAMPAIGN_ID
```

File: src/Symfony/Component/Notifier/Bridge/Iterable/README.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+```
+
+where:
+
+-   `API_KEY` is your Iterable API key
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iterable/README.md]:
```suggestion
 - `API_KEY` is your Iterable API key
```

CAMPAIGN_ID is missing in the explanation

File: src/Symfony/Component/Notifier/Bridge/Iterable/Tests/IterableOptionsTest.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+final class IterableOptionsTest extends TestCase
+{
+    public function testIterableOptionsWithSetters()
+    {
+        $options = (new IterableOptions())
+            ->recipientEmail('test@email.com')
+            ->recipientUserId('1234')
+            ->dataFields(['field' => 'value'])
+            ->sendAt('2021-05-05 01:01:00')
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iterable/Tests/IterableOptionsTest.php]: sendAt should receive a DateTimeInterface stead of a string

File: src/Symfony/Component/Notifier/Bridge/Iterable/Tests/IterableTransportFactoryTest.php [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+     */
+    public function createFactory(): TransportFactoryInterface
+    {
+        return new IterableTransportFactory();
+    }
+
+    public function createProvider(): iterable
+    {
+        yield ['iterable://host.test', 'iterable://token@host.test'];
+        yield ['iterable://host.test', 'iterable://token@host.test?campaign_id=testCampaignId'];
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iterable/Tests/IterableTransportFactoryTest.php]: The campaignID should be part of the result

File: src/Symfony/Component/Notifier/Bridge/Iterable/composer.json [hunk 7]
```diff
@@ -0,0 +1,30 @@
+{
+    "name": "symfony/iterable-notifier",
+    "type": "symfony-bridge",
+    "description": "Symfony Iterable Notifier Bridge",
+    "keywords": ["iterable", "notifier"],
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iterable/composer.json]: Please add "chat" as keyword

**Review by OskarStark — COMMENTED** [review]
Please add the package to the UnsupportedSchemeException and UnsupportedSchemeExceptionTest 

Thanks

**Comment by OskarStark** [comment]
And please target 🎯 branch 5.4

**Comment by balazscsaba2006** [comment]
@OskarStark I made all the changes to the code, thank you for reviewing

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php [hunk 8]
```diff
@@ -1878,7 +1878,7 @@ public function testIfNotifierTransportsAreKnownByFrameworkExtension()
     {
         $container = $this->createContainerFromFile('notifier');
 
-        foreach ((new Finder())->in(\dirname(__DIR__, 4).'/Component/Notifier/Bridge')->directories()->depth(0)->exclude('Mercure') as $bridgeDirectory) {
+        foreach ((new Finder())->in(\dirname(__DIR__, 4).'/Component/Notifier/Bridge')->directories()->depth(0)->exclude(['Mercure', 'Iterable']) as $bridgeDirectory) {
```
Comment (OskarStark) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php]:
```suggestion
        foreach ((new Finder())->in(\dirname(__DIR__, 4).'/Component/Notifier/Bridge')->directories()->depth(0)->exclude(['Mercure']) as $bridgeDirectory) {
```

Please remove and let the test fail for now, thanks

File: src/Symfony/Component/Notifier/Bridge/Iterable/IterableOptions.php [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+namespace Symfony\Component\Notifier\Bridge\Iterable;
+
+use Symfony\Component\Notifier\Message\MessageOptionsInterface;
+
+/**
+ * @author Balázs Csaba <csaba.balazs@lingoda.com>
+ *
+ * @see https://api.iterable.com/api/docs#push_target
+ */
+class IterableOptions implements MessageOptionsInterface
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iterable/IterableOptions.php]:
```suggestion
final class IterableOptions implements MessageOptionsInterface
```

File: src/Symfony/Component/Notifier/Bridge/Iterable/IterableOptions.php [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+{
+    /**
+     * @var array<mixed>
+     *
+     * @see https://api.iterable.com/api/docs#schema-push_target_body
+     */
+    protected $options;
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iterable/IterableOptions.php]:
```suggestion
    private $options;
```

**Comment by OskarStark** [comment]
👋 Hi @balazscsaba2006 

are you willing to finish this PR?

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php [hunk 11]
```diff
@@ -1878,7 +1878,7 @@ public function testIfNotifierTransportsAreKnownByFrameworkExtension()
     {
         $container = $this->createContainerFromFile('notifier');
 
-        foreach ((new Finder())->in(\dirname(__DIR__, 4).'/Component/Notifier/Bridge')->directories()->depth(0)->exclude('Mercure') as $bridgeDirectory) {
+        foreach ((new Finder())->in(\dirname(__DIR__, 4).'/Component/Notifier/Bridge')->directories()->depth(0)->exclude(['Mercure']) as $bridgeDirectory) {
```
Comment (OskarStark) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php:1881]: Why this change?

Comment (OskarStark) [reply src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php:1881]: Ah its based on https://github.com/symfony/symfony/pull/41713#discussion_r653378795

Can you please remove the `[]`? Thanks

**Comment by OskarStark** [comment]
Can you please add the new bridge to the `UnsupportedSchemeExceptionTest`? Thanks

File: src/Symfony/Component/Notifier/Bridge/Iterable/IterableTransport.php [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+        if ($message->getOptions() && !$message->getOptions() instanceof IterableOptions) {
+            throw new LogicException(sprintf('The "%s" transport only supports instances of "%s" for options.', __CLASS__, IterableOptions::class));
+        }
+
+        $endpoint = sprintf('https://%s', $this->getEndpoint());
+        $options = ($opts = $message->getOptions()) ? $opts->toArray() : [];
+
+        $campaignId = $options['campaignId'] ?? $this->campaignId;
+        if (null === $campaignId) {
+            throw new InvalidArgumentException(sprintf('The "%s" transport required the "campaignId" to be set.', __CLASS__));
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iterable/IterableTransport.php:77]: Is it really a common use case to not set a campaign_id on the DSN and set different ones on the ChatMessage via options?

If not, I would make the campaign_id required and not nullable and remove it from the IterableOptions

File: src/Symfony/Component/Notifier/Bridge/Iterable/IterableTransportFactory.php [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+        if ('iterable' !== $scheme) {
+            throw new UnsupportedSchemeException($dsn, 'iterable', $this->getSupportedSchemes());
+        }
+
+        $apiKey = $this->getUser($dsn);
+        $campaignId = $dsn->getOption('campaign_id');
+        $host = 'default' === $dsn->getHost() ? null : $dsn->getHost();
+        $port = $dsn->getPort();
+
+        return (new IterableTransport($apiKey, $campaignId, $this->client, $this->dispatcher))->setHost($host)->setPort($port);
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iterable/IterableTransportFactory.php:39]:
```suggestion
        if ('iterable' !== $scheme) {
            $apiKey = $this->getUser($dsn);
            $campaignId = $dsn->getOption('campaign_id');
            $host = 'default' === $dsn->getHost() ? null : $dsn->getHost();
            $port = $dsn->getPort();

            return (new IterableTransport($apiKey, $campaignId, $this->client, $this->dispatcher))->setHost($host)->setPort($port);    
        }
        
        throw new UnsupportedSchemeException($dsn, 'iterable', $this->getSupportedSchemes());
```

File: src/Symfony/Component/Notifier/Bridge/Iterable/README.md [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+## DSN example
+
+```
+ITERABLE_DSN=iterable://API_KEY@default?campaign_id=CAMPAIGN_ID
+```
+
+where:
+
+ - `API_KEY` is your Iterable API key
+ - `CAMPAIGN_ID` (optional) is the campaign's ID where to send messages (for e.g. push notifications) - The campaign ID can be set on the `IterableOptions` as well
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iterable/README.md:14]: This sentence needs to be adjusted after the decision in https://github.com/symfony/symfony/pull/41713/files#r680538202

File: src/Symfony/Component/Notifier/Bridge/Iterable/Tests/IterableTransportTest.php [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+        $client = new MockHttpClient(static function () use ($response): ResponseInterface {
+            return $response;
+        });
+
+        $transport = $this->createTransport($client, 'testCampaignId');
+        try {
+            $transport->send(new ChatMessage('testMessage'));
+        } catch (TransportException $exception) {
+            $this->fail('TransportException thrown');
+        }
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iterable/Tests/IterableTransportTest.php:96]:
```suggestion
        $this->expectException(TransportException::class);

        $transport->send(new ChatMessage('testMessage'));
```

Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Iterable/Tests/IterableTransportTest.php:96]: This is not needed.

It is better to `$sentMessage = $transport->send(....)` and then `assertSame($id, $sentMessage->getId())`

File: src/Symfony/Component/Notifier/Bridge/Iterable/README.md [hunk 16]
```diff
@@ -0,0 +1,21 @@
+# Iterable Notifier
+
+Provides [Iterable](https://iterable.com/) integration for Symfony Notifier.
```
Comment (jderusse) [inline src/Symfony/Component/Notifier/Bridge/Iterable/README.md:3]: ohh :-( when reading the title of this PR, I thought it was about a special provider that broadcast (or fallback) over a list of providers ..

File: src/Symfony/Component/Notifier/Bridge/Iterable/IterableOptions.php [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        $this->options['recipientEmail'] = $recipientEmail;
+
+        return $this;
+    }
+
+    /**
+     * UserId that was passed into the updateUser call.
+     */
+    public function recipientUserId(string $recipientUserId): self
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Notifier/Bridge/Iterable/IterableOptions.php:73]: please add `@return $this` annotations every where possible

**Comment by fabpot** [comment]
@balazscsaba2006 Are you still interested in finishing this PR?

**Comment by fabpot** [comment]
Closing as there is no activity anymore, feel free to reopen when you are ready. Thank you.
